src/arns.lua

```diff
@@ -395,24 +391,113 @@ function arns.getTokenCost(intendedAction)
 			yearsRemaining = arns.calculateYearsBetweenTimestamps(currentTimestamp, record.endTimestamp)
 		end
 		tokenCost = arns.calculateUndernameCost(baseFee, qty, record.type, yearsRemaining, demand.getDemandFactor())
+	-- TODO: move action map to constants and use it here
+	elseif intent == "Upgrade-Name" then
+		local record = arns.getRecord(name)
+		if not record then
+			error("Name is not registered")
+		end
+
+		if record.type == "permabuy" then
+			error("Name is already a permabuy")
+		end
+		tokenCost = arns.calculatePermabuyFee(baseFee, demand.getDemandFactor())
 	end
 	return tokenCost
 end
 
-function arns.assertValidIncreaseUndername(record, qty, currentTimestamp)
+--- Upgrades a leased record to a permabuy
+--- @param from string The address of the sender
+--- @param name string The name of the record
+--- @param currentTimestamp number The current timestamp
+--- @return table The upgraded record with name and record fields
+function arns.upgradeRecord(from, name, currentTimestamp)
+	local record = arns.getRecord(name)
 	if not record then
 		error("Name is not registered")
 	end
 
-	if
-		record.endTimestamp
+	if record.type == "permabuy" then
+		error("Record is already a permabuy")
+	end
+	if arns.recordExpired(record, currentTimestamp) then
+		error("Name is expired")
+	end
+
+	local baseFee = demand.baseFeeForNameLength(#name)
+	local demandFactor = demand.getDemandFactor()
+	local upgradeCost = arns.calculatePermabuyFee(baseFee, demandFactor)
+
+	if not utils.walletHasSufficientBalance(from, upgradeCost) then
+		error("Insufficient balance")
+	end
+
+	record.endTimestamp = nil
+	record.type = "permabuy"
+	record.purchasePrice = upgradeCost
+
+	balances.transfer(ao.id, from, upgradeCost)
+	demand.tallyNamePurchase(upgradeCost)
+
+	NameRegistry.records[name] = record
+	return {
+		name = name,
+		record = record,
+		totalUpgradeFee = upgradeCost,
+		baseRegistrationFee = baseFee,
+		remainingBalance = balances.getBalance(from),
```

Comment:
These two aren't as important to return since you can access them from the globals. ndb though